st7735/lcd_st7735.c

```diff
@@ -307,3 +307,125 @@ Result lcd_st7735_reset(St7735Context *ctx, bool hw) {
+// (unfortunately error-ridden) ST7735 datasheet for more details.
+//
+// NOTE1: Must be run after a HW or SW reset and before any CASET commands.
+// NOTE2: Does NOT always perform a reset before returning. State may be dirty.
+// Author: [Elliot Baptist](https://github.com/elliotb-lowrisc)
+
+Result lcd_st7735_check_offset(St7735Context *ctx, size_t *offset_x, size_t *offset_y) {
+  enum {
+    attempts = 3,
+    buf_len  = 4,
+  };
+
+  uint8_t buf[buf_len];
+  const uint8_t patterns[4] = {0xA8, 0xCC, 0xE0, 0x90};
+  uint8_t result;
+
+  for (unsigned iter = 0; iter < attempts; iter++) {
+    // Ensure CS line is de-asserted and output enabled ahead of any commands
+    ctx->parent.interface->gpio_write(ctx->parent.interface->handle, true, false);
+
+    // Select 18-bit pixel format. Affects writes only (reads always 18-bit).
+    // 18-bit pixel format (as per ST7735 datasheet):
+    //
+    //  MSB                                                                 LSB
+    //  R5 R4 R3 R2 R1 R0 -- -- G5 G4 G3 G2 G1 G0 -- -- B5 B4 B3 B2 B1 B0 -- --
+    // | First pixel byte      | Second pixel byte     | Last pixel byte       |
+    //
+    // Where "R5" is the first bit on the wire, and "--" bits are ignored.
+    write_command(ctx, ST7735_COLMOD);
+    ctx->parent.interface->gpio_write(ctx->parent.interface->handle, false, true);
+    uint8_t value = 0x06;
+    write_buffer(ctx, &value, sizeof(value));
+
+    // Write 4 lots (possibly lines) of 132 pixels into the frame buffer.
+    // Change the value being written every 132 pixels.
+    write_command(ctx, ST7735_RAMWR);
+    ctx->parent.interface->gpio_write(ctx->parent.interface->handle, false, true);
+    for (unsigned l = 0u; l < sizeof(patterns); l++) {
+      for (unsigned p = 0u; p < 132; p++) {
+        // 18-bit pixel value packed into 24-bit (3 bytes) payload.
+        // Two padding LSBs and 6 MSBs of real data per byte/channel.
+        buf[0] = patterns[l];  // red
+        buf[1] = patterns[l];  // green
+        buf[2] = patterns[l];  // blue
+        write_buffer(ctx, buf, 3);
+      }
+    }
+
+    // Read back a pixel from the start of the second, third, and fourth lines
+    // of the external frame buffer to determine whether the ST7735 controller
+    // in the LCD assembly is configured for 128x160 or 132x162 (by GM pads).
+    // Pixels are always read back in 18-bit format, regardless of COLMOD.
+    result = 0;
+    for (uint8_t l = 1u; l < sizeof(patterns); l++) {
+      write_command(ctx, ST7735_RASET);
+
+      buf[0] = l >> 8;
+      buf[1] = l;
+      buf[2] = 99 >> 8;
+      buf[3] = 99;
+      ctx->parent.interface->gpio_write(ctx->parent.interface->handle, false, true);
+      write_buffer(ctx, buf, 4);
+
+      write_command(ctx, ST7735_RAMRD);
+      // Read 1 dummy byte and 3 actual bytes (offset by a dummy clock cycle)
+      ctx->parent.interface->spi_read(ctx->parent.interface->handle, buf, 4);
```

Comment:
What happens if spi_read is null for example for old bitstreams that don't have a SPI read port?

Reply:
I added a check in the beginning of the function.